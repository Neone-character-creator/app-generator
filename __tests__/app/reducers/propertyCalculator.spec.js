const PropertyCalculator = require("../../../project-template/src/main/resources/scripts/reducer/propertyCalculator").default;

describe("property calculator", function(){
    let calculator;
    let model;
    let state;
    beforeEach(()=>{
        calculator = new PropertyCalculator();
        model = {
            character: {
                properties: {
                    name: {
                        type: "string"
                    },
                    age: {
                        type: "number"
                    },
                    nextAge: {
                        type: "number",
                        derivedFrom: [
                            "#$state.character.age + 1"
                        ]
                    },
                    doubleAge: {
                        type: "number",
                        derivedFrom: [
                            "#$state.character.nextAge * 2"
                        ]
                    }
                }
            }
        };
        state = {
            character: {
                name: "name",
                age: 1
            }
        }
    });
    it("return the number of iterations it performs", function(){
        expect(calculator.calculate({type: "character", modelDefinition: model.character}, state, ["character"], {})).toBe(2);
    });
    it("throws an error if more than 2 iterations are performed", function(){
        model.character = {
            properties: {
                a: {
                    type: "number",
                    derivedFrom: ["#$state.character.b + 1"]
                },
                b: {
                    type: "number",
                    derivedFrom: ["#$state.character.a + 1"]
                }
            }
        };
        state.character.a = 1;
        state.character.b = 1;
        expect(()=>{
            calculator.calculate({type: "character", modelDefinition: model.character}, state, ["character"], {})
        }).toThrowErrorMatchingSnapshot();
    });
    it("calculates base values after derived properties", function(){
        calculator.calculate({type: "character", modelDefinition: model.character}, state, ["character"], {});
        expect(state.character.nextAge).toEqual(2)
        expect(state.character.doubleAge).toEqual(4)
    })
});