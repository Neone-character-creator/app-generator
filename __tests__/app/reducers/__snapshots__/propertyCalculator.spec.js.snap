// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`property calculator throws an error if more than 2 iterations are performed 1`] = `
"Calculated property evaluation took more than 3 iterations. This likely is caused by a circular reference, where properties are causing each other to change in an infinite loop
The following properties were modified in the last evaluation cycle: character[a] = 6
character[b] = 7"
`;
